Mathlib/Topology/Germ.lean

```diff
@@ -0,0 +1,190 @@
+shared by all representatives at the given point. -/
+def value {X α : Type*} [TopologicalSpace X] {x : X} (φ : Germ (𝓝 x) α) : α :=
+  Quotient.liftOn' φ (fun f ↦ f x) fun f g h ↦ by dsimp only; rw [Eventually.self_of_nhds h]
+
+theorem value_smul {α β : Type*} [SMul α β] (φ : Germ (𝓝 x) α)
+    (ψ : Germ (𝓝 x) β) : (φ • ψ).value = φ.value • ψ.value :=
+  Germ.inductionOn φ fun _ ↦ Germ.inductionOn ψ fun _ ↦ rfl
+
+/-- The map `Germ (𝓝 x) E → E` as a monoid homeomorphism -/
+@[to_additive "The map `Germ (𝓝 x) E → E` as an additive monoid homeomorphism"]
+def valueMulHom {X E : Type*} [Monoid E] [TopologicalSpace X] {x : X} : Germ (𝓝 x) E →* E where
+  toFun := Filter.Germ.value
+  map_one' := rfl
+  map_mul' φ ψ := Germ.inductionOn φ fun _ ↦ Germ.inductionOn ψ fun _ ↦ rfl
+
+/-- The map `Germ (𝓝 x) E → E` into a `𝕜`-module `E` as a `𝕜`-linear map -/
+def valueₗ {X 𝕜 E : Type*} [Semiring 𝕜] [AddCommMonoid E] [Module 𝕜 E] [TopologicalSpace X]
+    {x : X} : Germ (𝓝 x) E →ₗ[𝕜] E :=
+  { Filter.Germ.valueAddHom with map_smul' := fun _ φ ↦ Germ.inductionOn φ fun _ ↦ rfl }
+
+/-- The map `Germ (𝓝 x) E → E` as a ring homeomorphism -/
+def valueRingHom {X E : Type*} [Semiring E] [TopologicalSpace X] {x : X} : Germ (𝓝 x) E →+* E :=
+  { Filter.Germ.valueMulHom, Filter.Germ.valueAddHom with }
+
+/-- The map `Germ (𝓝 x) E → E` as a monotone ring homeomorphism -/
+def valueOrderRingHom {X E : Type*} [OrderedSemiring E] [TopologicalSpace X] {x : X} :
+    Germ (𝓝 x) E →+*o E where
+  __ := Filter.Germ.valueRingHom
+  monotone' := fun φ ψ ↦
+  Germ.inductionOn φ fun _ ↦ Germ.inductionOn ψ fun _ h ↦ h.self_of_nhds
+
+end Filter.Germ
+
+/-- The inclusion `S → R` of a subring, as an ordered ring homomorphism. -/
+-- xxx: OrderedRing has no morphisms, OrderedRingHom no subtypes -> which file is a good place?
+def _root_.Subring.orderedSubtype {R} [OrderedRing R] (s : Subring R) : s →+*o R :=
+  { s.subtype with monotone' := fun _ _ h ↦ h }
+
+section RestrictGermPredicate
+/-- Given a predicate on germs `P : Π x : X, germ (𝓝 x) Y → Prop` and `A : set X`,
+build a new predicate on germs `restrict_germ_predicate P A` such that
+`(∀ x, restrict_germ_predicate P A x f) ↔ ∀ᶠ x near A, P x f`, see
+`forall_restrict_germ_predicate_iff` for this equivalence. -/
+def RestrictGermPredicate (P : ∀ x : X, Germ (𝓝 x) Y → Prop)
+    (A : Set X) : ∀ x : X, Germ (𝓝 x) Y → Prop := fun x φ ↦
+  Quotient.liftOn' φ (fun f ↦ x ∈ A → ∀ᶠ y in 𝓝 x, P y f)
+    haveI : ∀ f f' : X → Y, f =ᶠ[𝓝 x] f' → (∀ᶠ y in 𝓝 x, P y f) → ∀ᶠ y in 𝓝 x, P y f' := by
+      intro f f' hff' hf
+      apply (hf.and <| Eventually.eventually_nhds hff').mono
+      rintro y ⟨hy, hy'⟩
+      rwa [Germ.coe_eq.mpr (EventuallyEq.symm hy')]
+    fun f f' hff' ↦ propext <| forall_congr' fun _ ↦ ⟨this f f' hff', this f' f hff'.symm⟩
+
+theorem Filter.Eventually.germ_congr
+    {P : Germ (𝓝 x) Y → Prop} (hf : P f) (h : ∀ᶠ z in 𝓝 x, g z = f z) : P g := by
+  convert hf using 1
+  exact Germ.coe_eq.mpr h
```

Comment:
This looks a bit weird to me, as it seems like a very specialized combination of already existing facts. Is it useful for dot notation? Could you link some uses in the sphere eversion project so I see where it's useful?

Reply:
All uses of `germ_congr{,_set}` are in [InductiveConstruction](https://github.com/leanprover-community/sphere-eversion/blob/master/SphereEversion/InductiveConstructions.lean). There are four places overall; it might be easiest to just use the search function.

Reply:
I looked now: the one use of germ_congr is easy to inline. I would suggest I remove `germ_congr` and keep `germ_congr_set`. What do you think?

Reply:
I inlined `germ_congr` into `germ_congr_set`.